Add PHP parser based on PEG.js grammar

nylen opened the pull request:
There is a lot left to be done here, but this PR takes steps towards unifying our parsing logic across JS and PHP by using the same PEG grammar in both places.

There is a library for this that we can use: [`php-pegjs`](https://github.com/Nordth/php-pegjs).  It's a PEG.js plugin that generates a parser in PHP instead of JavaScript.  However, there are lots of bits of JavaScript code interspersed in our PEG to make it work, and unfortunately, `php-pegjs` expects those to be written in PHP only.

So, this PR also points to my fork of `php-pegjs` which allows these code blocks to be specified in both PHP and JavaScript (see https://github.com/Nordth/php-pegjs/pull/5).

Once finished, this will fix #1086 and hopefully also other issues such as #882.

Remaining items:

- [x] PHP 5.2 support: Remove namespaces in generated `php-pegjs` output; allow specifying a function prefix instead. Eliminate use of closures.
- [x] Actually use the generated parser in the plugin
- [x] Add PHPUnit tests for parser output using our [existing HTML fixtures](https://github.com/WordPress/gutenberg/blob/a228bc0/blocks/test/fixtures/README.md).
- [x] Require `php-pegjs` as a normal, stable dependency, by getting the needed changes merged into the main repo or a fork.

It would also be nice to upgrade `php-pegjs` to support newer versions of `pegjs` - while downgrading to `pegjs` 0.8.x causes the [`pegjs-loader` peer dependency](https://github.com/eploko/pegjs-loader/blob/v0.5.1/package.json#L28) not to be met, in the actual code [this will work fine](https://github.com/eploko/pegjs-loader/commit/7a65f1c447e8f7e94b0dfbdc942c62dbb7f760f2).

nylen committed:
Introduce `php-pegjs`

Add `php-pegjs` and downgrade `pegjs` version accordingly

Add script to generate PHP parser

Update PEG grammar with PHP and JS code

Run syntax check against php-pegjs parser

Exclude generated PHP parser from phpcs checks

Fix lint errors

dmsnell commented:
Thanks @nylen!

This is something I had also planned on looping back around to. I'm thinking it may be possible to end up with a grammar that has little-to-no JS _or_ PHP.

That shouldn't stop this from moving forward, but I think that we can be carefully plan out what the parser should be parsing and ignoring and automatically return the right data structures. For example, I was thinking of ripping out the `keyValue` function anyway if I can ignore the part of the parse we don't want.

westonruter commented:
Awesome. I'm excited about this.

Another problem with the generated PHP parser in 5.2 is that it using closures.

westonruter commented on .travis.yml:
```diff
@@ -50,8 +50,14 @@ before_script:
       npm install || exit 1
       npm run build || exit 1
+      # Syntax check for php-pegjs parser
+      node bin/create-php-parser.js || exit 1
```
Should this be added to the NPM `build` and `dev` scripts?

nylen replied:
I think not yet - instead, we need to make sure we run our existing parsing tests against the PHP parser as well.

aduth commented on blocks/api/post.pegjs:
```diff
@@ -21,23 +25,59 @@ WP_Block
+  {
+    /** <?php
+    return array(
+      'blockType'  => $blockType,
```
Note: With #974, we'll need to update naming here to `blockName`

dmsnell commented:
I spoke with @westonruter today about this and I'm really wondering if this is the right path. I'm pretty sure that the generated parser for PHP will be very sub-par on performance and the library doesn't appear to have been touched in over two years.

@nylen what do you think about simply having a separate parser file for PHP. Copying this for now and replacing the JS bits with PHP would seem reasonable, but I think that in the end we might be better served by a hand-written parser in PHP that implements the PEG spec simply on account of the different PHP semantics.

nylen commented:
> @nylen what do you think about simply having a separate parser file for PHP

I missed this conversation, but I'm not a fan of having separate parsing logic, as it guarantees that this critical piece of logic will continue to diverge between client and server.  See #1190 and #1213 for recent examples, and note that we still can't or don't test the server-side parser for these cases or to nearly the same extent as we do the client-side parser.

> I'm pretty sure that the generated parser for PHP will be very sub-par on performance

What are your specific concerns here?  PEG parsers are supposed to be quite performant.

> and the library doesn't appear to have been touched in over two years

This is true; it seems likely that we will have to fork it for our needs.

dmsnell commented:
> What are your specific concerns here? PEG parsers are supposed to be quite performant.

My concern is how the parser generator will generate the PHP and the overhead it brings with it. My fears should not be trusted until we know it's a problem. On the other hand, it's hard to establish a baseline. How many `ms` should it take to parse a document?

> as it guarantees that this critical piece of logic will continue to diverge between client and server

I see your point though it's my hope that we will always end up with more than one. The spec parser which is _clear_ to read and implementations that meet the spec but are faster. Already I've had some decisions to tradeoff readability against speed _within the PEG_

nylen committed:
Change php-pegjs to phpegjs (forked version)

nylen committed:
Add parser output to full-content test fixtures

nylen committed:
Run tests against PHP parser during Travis CI builds

Or using the following commands (TODO - add a npm script for this?):

```
bin/create-php-parser.js
RUN_PARSER_TESTS=1 phpunit
```

nylen committed:
Update parser fixture files after rebase

nylen commented:
I've rebased and updated this PR to use [`phpegjs`](https://github.com/nylen/phpegjs), my fork of `php-pegjs`.  It has WIP support for PEG.js 0.10.0 (the latest version, which we have been using in this repository).

There were some [significant changes](https://github.com/nylen/phpegjs/commit/fd97dfc0f975c5ca8124490e7c026302cbb31d58) required to support the new version of PEG.js, and there are still some bugs, as evidenced by the failing PHP parser tests (newly introduced in this PR).

However, this is good progress all the same, and I still think this task is doable.

nylen committed:
Decode block JSON attributes as PHP arrays, not StdClass objects

nylen committed:
Remove error logging and clarify variable name

nylen committed:
Upgrade `phpegjs` to 1.0.0-beta2

Fix all the things (except PHP 5.2 support).

nylen committed:
Use a different method of skipping the PHP parser tests

Otherwise PHPUnit 4.5.0 (used with PHP 5.6 Travis build) complains
about a suite with no tests in it.

nylen committed:
Make phpcs happy again

nylen commented:
One step closer: the generated PHP parser now has a bunch of tests based on our [existing fixtures](https://github.com/WordPress/gutenberg/blob/a228bc0/blocks/test/fixtures/README.md).  The logic here has also been rebased to support JSON-style attributes (#1213).

Still left to do: PHP 5.2 support; actually use the new parser for `do_blocks` server-side parsing and rendering.

paulwilde commented:
Honest question - Why should Gutenberg support a 11 year old version of PHP? Surely by the time Gutenberg is merged into core (Late 2017/Early 2018) it's time to think about updating the minimum requirements for PHP.

Gutenberg is a nice forward step in not making WordPress a modern piece of software, so why should the PHP side of things be left behind?

nylen commented:
> Honest question - Why should Gutenberg support a 11 year old version of PHP?

It would be nice if we didn't have to support PHP 5.2.  However, I think this decision should be made along with the rest of core, and there are ongoing discussions about this.

Quoting from [recent discussion in the #core-php Slack channel](https://wordpress.slack.com/archives/C60K3MP2Q/p1498656688342571):

> There will be a summary posted from the summit discussion within the next week or so (notes are being collected from the note takers currently). The consensus was to do as much as we can to encourage upgrading to see if we can get the percentage of users on old PHP versions down, and revisit in a little while. #41191 and #40934 were the first steps. Would be great to get further feedback on that post once it happens.

As far as supporting PHP 5.2 in this PR, the work is already mostly done, via [this commit](https://github.com/nylen/phpegjs/commit/7d2fdb13166e3441ed26f38ead7110713a8912ef).

nylen committed:
Upgrade `phpegjs` to 1.0.0-beta3

Adds support for generating a parser compatible with PHP 5.2.

nylen committed:
Do not use namespaces in generated PHP parser

nylen committed:
Add generated PHP parser to repository

This will simplify the process of loading the plugin.  Otherwise we'd
have to add this step to `npm run dev` and a couple of other places.

nylen committed:
During Travis build, make sure PHP parser is up to date

nylen committed:
Modify the PHP parser (this should fail the build)

nylen committed:
Fix a few test fixtures that needed to change after rebase

nylen committed:
Revert "Modify the PHP parser (this should fail the build)"

This reverts commit 57389f49de1bce57bf97012df4d207bf2fffdea3.

nylen committed:
Use PHP PEG parser in plugin

nylen commented:
This is ready for review/merge.

nylen commented:
Merging to get this and the breaking change layered on top of it (#1593) into a release as soon as possible.

aduth commented:
These changes define `pegjs` as a duplicate dependency between `dependencies` and `devDependencies`, with a warning on `npm install`:

```
npm WARN The package pegjs is included as both a dev and production dependency.
```

nylen commented:
>These changes define `pegjs` as a duplicate dependency between `dependencies` and `devDependencies`, with a warning on `npm install`

Fixed in #1708.
